packages/decoder/lib/decoders.ts

```diff
@@ -133,6 +135,124 @@ export class ProjectDecoder {
+    this.contractsAndContexts = [
+      ...this.contractsAndContexts,
+      ...contractsAndContexts
+    ];
+
+    //everything up till now has been pretty straightforward merges.
+    //but allocations are a bit more complicated.
+    //some of them can be merged straightforwardly, some can't.
+    //we'll start with the straightforward ones.
+    const abiAllocations = AbiData.Allocate.getAbiAllocations(userDefinedTypes);
+    Object.assign(this.allocations.abi, abiAllocations);
+    const storageAllocations = Storage.Allocate.getStorageAllocations(
+      userDefinedTypesByCompilation
+    );
+    Object.assign(this.allocations.storage, storageAllocations);
+    const stateAllocations = Storage.Allocate.getStateAllocations(
+      allocationInfo,
+      referenceDeclarations,
+      userDefinedTypes,
+      storageAllocations //only need the ones from the new compilations
+    );
+    Object.assign(this.allocations.state, stateAllocations);
+
+    //now we have calldata allocations.  merging these is still mostly straightforward,
+    //but slightly less so.
+    const calldataAllocations = AbiData.Allocate.getCalldataAllocations(
+      allocationInfo,
+      referenceDeclarations,
+      userDefinedTypes,
+      abiAllocations //only need the ones from the new compilations
+    );
+    //merge constructor and function allocations separately
+    Object.assign(
+      this.allocations.calldata.constructorAllocations,
+      calldataAllocations.constructorAllocations
+    );
+    Object.assign(
+      this.allocations.calldata.functionAllocations,
+      calldataAllocations.functionAllocations
+    );
+
+    //finally, redo the allocations for returndata and eventdata.
+    //attempting to perform a merge on these is too complicated, so we
+    //won't try; we'll just recompute.
+    this.allocations.returndata = AbiData.Allocate.getReturndataAllocations(
+      allocationInfo,
+      referenceDeclarations,
+      userDefinedTypes,
+      this.allocations.abi //we're doing this for merged result, so use merged input!
+    );
+    this.allocations.event = AbiData.Allocate.getEventAllocations(
+      allocationInfo,
+      referenceDeclarations,
+      userDefinedTypes,
+      this.allocations.abi //we're doing this for merged result, so use merged input!
+    );
+  }
+
+  /**
+   * **This function is asynchronous.**
+   *
+   * Adds additional compilations to the decoder like [[addCompilations]],
+   * but allows it to be specified in more general forms.
+   *
+   * @param projectInfo Information about the additional compilations or
+   * contracts to be decoded.  This may come in several forms; see the type
+   * documentation for more information.  If passing in `{ compilations: ... }`,
+   * take care that the compilations have different IDs from others passed in
+   * so far.  If passed in in another form, an ID will be assigned automatically,
+   * which should generally avoid any collisions.
```

Comment:
This sounds like it would be a difficult thing for users to debug. Is there a way to give feedback that the ID exists? Am I missing something?

Reply:
OK, I'll add that!

Reply:
OK, I've added this.  In addition to checking for collisions on adding new ones, I also added checks for repeats among the ones you're adding, and among the ones you start the decoder with.  I also added tests of this.

Although, hm -- I just realized I didn't add a check for repeats when starting *en*coder... do you think that should be added too?

Reply:
Sure, it makes sense to be thorough with the feedback.